### PR TITLE
Integration tests for livelog and fix for s3 upload

### PR DESCRIPTION
**Pull request by jonasfj**

### CODE REVIEW

**Comment by jonasfj** [comment]
Now this also fixes empty artifact upload, issue #375

File: worker/workertest/superseded_test.go [hunk 1]
```diff
@@ -88,42 +35,101 @@ func TestSuperseded(t *testing.T) {
...
+				// Get taskId from which this is being called
+				taskID := r.URL.Query().Get("taskId")
+
+				// Create array of superseding taskIds, always start with its own taskId
+				supersedes := []string{taskID}
+
+				// If we haven't made a superseded task yet we do that now
+				if supersededTaskID == "" {
+					supersededTaskID = taskID
+					// Make first other taskId be superseding this taskId
```
Comment (walac) [inline worker/workertest/superseded_test.go]: I didn't understand this sentence

File: test/nativetest/artifacts_posix_test.go [hunk 2]
```diff
@@ -11,144 +11,144 @@ import (
...
+				Success: false,
+				Payload: `{
+					"command": ["true"],
+					"env": {},
+					"maxRunTime": "10 minutes",
+					"artifacts": [
+						{
+							"name": "public/result.txt",
+							"type": "file",
+							"path": "no-such-file-` + filename + `.txt"
```
Comment (walac) [inline test/nativetest/artifacts_posix_test.go]: Where is the variable filename defined?

Comment (jonasfj) [reply test/nativetest/artifacts_posix_test.go]: Probably higher up...this is just a mechanically refactored test case.. see the commit

File: test/scripttest/logging_test.go [hunk 3]
```diff
@@ -57,47 +67,177 @@ func TestLogging(t *testing.T) {
...
 		Setup: func(t *testing.T, env workertest.Environment) func() {
-			// Create test server with magic-words, we'll grep for later
-			s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Create test-server with magic-words, we'll grep for later
+			var wg sync.WaitGroup
+			s = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				taskID := r.URL.Query().Get("taskId")
+				msg := messages[r.URL.Query().Get("msgSize")]
+				behavior := r.URL.Query().Get("behavior")
+				reply := replies[r.URL.Query().Get("replySize")]
```
Comment (walac) [inline test/scripttest/logging_test.go]: Are `msg` and `reply` buffers or buffer sizes?

File: test/scripttest/logging_test.go [hunk 4]
```diff
@@ -57,47 +67,177 @@ func TestLogging(t *testing.T) {
...
+					n, rerr := res.Body.Read(b)
+					require.NoError(t, rerr, "failed to read body")
+					data = append(data, b[0:n]...)
+					if bytes.Contains(data, []byte(msg)) {
+						debug("Read special message from livelog")
+						break
+					}
+				}
+
+				// Signal that eveyrthing went well, and return reply.
```
Comment (walac) [inline test/scripttest/logging_test.go]: nit: "everything"

File: worker/workertest/workertest.go [hunk 5]
```diff
@@ -299,31 +299,35 @@ func (c Case) testWithQueue(t *testing.T, q *tcqueue.Queue, l fakequeue.Listener
...
-		}(taskIDs[i])
+		tdefs[i].Scopes = task.Scopes
+		tdefs[i].Metadata.Name = title
+		tdefs[i].Metadata.Description = "Task from taskcluster-worker integration tests"
+		tdefs[i].Metadata.Source = "https://github.com/taskcluster/taskcluster-worker/tree/master/worker/workertest/workertest.go"
+		tdefs[i].Metadata.Owner = "jonasfj@mozilla.com"
+	}
+
+	// Create task asynchronously with a limit of 10 concurrent calls to make
+	// that it goes fairly faster, when there is a lot tasks.
```
Comment (walac) [inline worker/workertest/workertest.go]: nit: "a not of"

File: worker/workertest/workertest.go [hunk 6]
```diff
@@ -299,31 +299,35 @@ func (c Case) testWithQueue(t *testing.T, q *tcqueue.Queue, l fakequeue.Listener
...
-			require.NoError(t, terr, "Failed to create task: %s", title)
-		}(taskIDs[i])
+		tdefs[i].Scopes = task.Scopes
+		tdefs[i].Metadata.Name = title
+		tdefs[i].Metadata.Description = "Task from taskcluster-worker integration tests"
+		tdefs[i].Metadata.Source = "https://github.com/taskcluster/taskcluster-worker/tree/master/worker/workertest/workertest.go"
+		tdefs[i].Metadata.Owner = "jonasfj@mozilla.com"
+	}
+
+	// Create task asynchronously with a limit of 10 concurrent calls to make
```
Comment (walac) [inline worker/workertest/workertest.go]: nit: "task**s**"

File: worker/workertest/fakequeue/fakequeue.go [hunk 7]
```diff
@@ -529,6 +529,16 @@ func (q *FakeQueue) internalPutArtifact(taskID string, runID int, name string, p
 	// Real S3/azure is slow, let's add a bit of jitter to get some intermittent bugs
 	time.Sleep(time.Duration(rand.Intn(15)) * time.Millisecond) // sleep 0 - 15ms
 
+	// S3 always requires a content-length, if it's not present we have bug
```
Comment (walac) [inline worker/workertest/fakequeue/fakequeue.go]: nit: "have **a** bug"

File: test/scripttest/logging_test.go (hunk 3 above)
Comment (jonasfj) [reply test/scripttest/logging_test.go]: They are strings... `messages` is a mapping from _string representing size_ to _message as string_.

**Comment by jonasfj** [comment]
@walac, I added a fix for the broken tests in master... due to missing test fixes from the `/` fixes...
A previously commit in this sequence fixes a bug a that hid CI issues in one of the test configurations.

I also rebased and merged a few of the commits... So each commit does one thing.

**Pull request merged**